File: python/nedb/store.py
# SPDX-FileCopyrightText: 2026 INTERCHAINED LLC
# SPDX-License-Identifier: BUSL-1.1
# aof-db · a distribution of NEDB · © 2026 INTERCHAINED LLC × Eth-Interchained × Vex (Claude Opus 5)

"""
nedb.store — MVCC versioned key/value store with time-travel.

Each key maps to an ascending list of (seq, value) versions. Reads at HEAD return
the newest version; reads `as_of=N` return the newest version with seq <= N. This
gives snapshot isolation (readers never block writers) and time-travel for free,
because the version history IS the operation log's effect, replayed.
"""
from __future__ import annotations

import bisect
from typing import Any, Dict, List, Optional, Tuple

TOMB = object()  # tombstone marker for deletes


class MVCCStore:
    def __init__(self) -> None:
        self._v: Dict[str, List[Tuple[int, Any]]] = {}
        self._seqs: Dict[str, List[int]] = {}  # parallel seq list for bisect

    def put(self, key: str, value: Any, seq: int) -> None:
        self._v.setdefault(key, []).append((seq, value))
        self._seqs.setdefault(key, []).append(seq)

    def delete(self, key: str, seq: int) -> None:
        self._v.setdefault(key, []).append((seq, TOMB))
        self._seqs.setdefault(key, []).append(seq)

    def get(self, key: str, as_of: Optional[int] = None) -> Optional[Any]:
        chain = self._v.get(key)
        if not chain:
            return None
        if as_of is None:
            _, val = chain[-1]
        else:
            i = bisect.bisect_right(self._seqs[key], as_of) - 1
            if i < 0:
                return None
            _, val = chain[i]
        return None if val is TOMB else val

    def keys(self, prefix: str = "", as_of: Optional[int] = None) -> List[str]:
        # Lock-free read: take an atomic snapshot of the key set, then resolve
        # versions by `as_of`. Under the concurrent Sequencer a single committer
        # thread may add a NEW key while we snapshot; CPython raises RuntimeError
        # ("dict changed size during iteration") only during construction, so we
        # retry the snapshot (collisions are microsecond-rare). Reading the per-key
        # version chain itself is always safe — chains are append-only.
        snap: List[str]
        for _ in range(128):
            try:
                snap = list(self._v.keys())
                break
            except RuntimeError:
                continue
        else:
            snap = list(self._v.keys())
        out = []
        for k in snap:
            if prefix and not k.startswith(prefix):
                continue
            if self.get(k, as_of) is not None:
                out.append(k)
        return out

    def snapshot(self, prefix: str = "", as_of: Optional[int] = None) -> Dict[str, Any]:
        return {k: self.get(k, as_of) for k in self.keys(prefix, as_of)}

File: rust/crates/nedb-core/src/store.rs
// SPDX-FileCopyrightText: 2026 INTERCHAINED LLC
// SPDX-License-Identifier: BUSL-1.1
// aof-db · a distribution of NEDB · © 2026 INTERCHAINED LLC × Eth-Interchained × Vex (Claude Opus 5)

//! MVCC versioned store with time-travel. Each key holds an ascending vector of
//! `(seq, Option<Value>)` versions (`None` is a tombstone). Reads at HEAD take the
//! last version; `as_of` reads take the newest version with `seq <= as_of`.

use std::collections::HashMap;

use serde_json::Value;

#[derive(Default)]
pub struct MvccStore {
    v: HashMap<String, Vec<(u64, Option<Value>)>>,
}

impl MvccStore {
    pub fn put(&mut self, key: &str, val: Value, seq: u64) {
        self.v.entry(key.to_string()).or_default().push((seq, Some(val)));
    }

    pub fn delete(&mut self, key: &str, seq: u64) {
        self.v.entry(key.to_string()).or_default().push((seq, None));
    }

    pub fn get(&self, key: &str, as_of: Option<u64>) -> Option<&Value> {
        let chain = self.v.get(key)?;
        let idx = match as_of {
            None => chain.len() - 1,
            Some(s) => {
                // versions are appended in ascending seq order → binary search
                let mut lo = 0usize;
                let mut hi = chain.len();
                while lo < hi {
                    let mid = (lo + hi) / 2;
                    if chain[mid].0 <= s {
                        lo = mid + 1;
                    } else {
                        hi = mid;
                    }
                }
                if lo == 0 {
                    return None;
                }
                lo - 1
            }
        };
        chain[idx].1.as_ref()
    }

    pub fn keys(&self, prefix: &str) -> Vec<&String> {
        self.v
            .keys()
            .filter(|k| k.starts_with(prefix))
            .filter(|k| self.get(k, None).is_some())
            .collect()
    }
}
